### PR TITLE
metrics: use `dogstatsd-ruby` v5.x single thread mode when available, fallback on v4.x otherwise

### PR DESCRIPTION
**Pull request by remeh**
Since version 5.2.0 of dogstatsd-ruby, a new option `single_thread` avoid to spawn a companion thread to flush the metrics, which was causing error in some circumstances (in some applications using `fork`s).

This PR is:
  - using a normal dogstatsd instance if dogstatsd-ruby gem version is `< 5.0`
  - warning if dogstatsd-ruby gem version is `>= 5.0, < 5.2`
  - using the single_thread mode if dogstatsd-ruby gem version is `>= 5.2`

### CODE REVIEW

File: lib/ddtrace/metrics.rb
Comment (ivoanjo) [inline lib/ddtrace/metrics.rb]: Minor: Perhaps this can be marked `private`?

File: lib/ddtrace/metrics.rb
Comment (ivoanjo) [inline lib/ddtrace/metrics.rb:252]: ⬆️  You forgot to update the `#ignored_statsd_warning` method that still recommends users install ~> 4.0 🤣

File: lib/ddtrace/metrics.rb
Comment (ivoanjo) [inline lib/ddtrace/metrics.rb:271]: Is this "only warn but let users go ahead" behavior what we still want?

Previously, I could see how if users had previously migrated to directly use the 5.0 APIs forcing them to use a 4.x release would be annoying.

But now that there's a compatible 5.x release, should dd-trace-rb outright refuse to work with `dogstatsd-ruby` versions 5.0.0, 5.0.1, and 5.1.0? That way customers would never accidentally run into the issue anymore -- they either have a working dogstatsd version OR we disable the dogstatsd usage until they either upgrade to 5.2 or downgrade to 4.x.

File: lib/ddtrace/metrics.rb
Comment (remeh) [reply lib/ddtrace/metrics.rb:252]: Damn, thanks!

File: lib/ddtrace/metrics.rb
Comment (remeh) [reply lib/ddtrace/metrics.rb:272]: I would indeed prefer this behavior, I don't know if it is a common usage or if it is already something done in dd-trace-rb for another gem?

There is scenario where using 5.0, 5.0.1 or 5.1.0 won't lead to any issue, however, I don't see any reason to stick with these versions neither.

Comment (ivoanjo) [reply lib/ddtrace/metrics.rb:272]: I'd say so: on this very file, you'll notice that if you have a too old version of dogstatsd, we already just refuse to use it:

https://github.com/DataDog/dd-trace-rb/blob/461a627641755f4338e1e7dbb93d366d4534d968/lib/ddtrace/metrics.rb#L16-L23

Furthermore, we also [detect a few incompatible versions of gems and gracefully degrade the profiler](https://github.com/DataDog/dd-trace-rb/blob/master/lib/ddtrace/profiling/ext/cpu.rb#L43).

> There is scenario where using 5.0, 5.0.1 or 5.1.0 won't lead to any issue, however, I don't see any reason to stick with these versions neither.

Indeed! I don't think there's any good reason for customers to use any of those versions anymore, so I would suggest we go the safe way :)

File: lib/ddtrace/metrics.rb
Comment (ivoanjo) [inline lib/ddtrace/metrics.rb:283]: Should this be `gem 'dogstatsd-ruby', '~> 5.2'` (also on the `GettingStarted.md`)?

Comment (remeh) [reply lib/ddtrace/metrics.rb:283]: phew, yes thanks. (sanity check: it will still upgrade to 5.3, 5.4, ..., ?)

File: lib/ddtrace/metrics.rb
Comment (remeh) [reply lib/ddtrace/metrics.rb:272]: You're right.
Let me refactor this.

Comment (remeh) [reply lib/ddtrace/metrics.rb]: Should be addressed

File: lib/ddtrace/metrics.rb
Comment (ivoanjo) [reply lib/ddtrace/metrics.rb]: Precisely -- you can see in the docs that apparently nobody calls this the twiddle-wakka: https://guides.rubygems.org/patterns/#pessimistic-version-constraint (lol and I don't either)

**Review by ivoanjo — APPROVED** [review]
Thanks a lot, it looks great to me! I'm guessing this is going to conflict with #1573 so it may be worth waiting until that one's merged later today :)

File: lib/ddtrace/metrics.rb
Comment (ericmustin) [reply lib/ddtrace/metrics.rb]: 👍 i think this approach makes sense as well

**Comment by marcotc** [comment]
😅 I had already from Monday: https://github.com/DataDog/dd-trace-rb/pull/1576/files :hide:

**Comment by remeh** [comment]
@marco whoops!

I'll close my PR and let you move forward with this then! Thanks having worked on this!

**Comment by ivoanjo** [comment]
Ooops x2, I missed that one as well 😰

**Pull request closed without merging**
